stable/cluster-autoscaler/templates/_helpers.tpl

```diff
@@ -76,9 +76,9 @@ Return the appropriate apiVersion for podsecuritypolicy.
 Return the service account name used by the pod.
 */}}
 {{- define "serviceaccount.name" -}}
-{{- if and .Values.rbac.create .Values.rbac.serviceAccount.create -}}
```

Comment:
Can we do this instead, as in `helm create cluster-autoscaler`?
```
{{/*
Create the name of the service account to use
*/}}
{{- define "cluster-autoscaler.serviceAccountName" -}}
{{- if .Values.serviceAccount.create -}}
    {{ default (include "cluster-autoscaler.fullname" .) .Values.serviceAccount.name }}
{{- else -}}
    {{ default "default" .Values.serviceAccount.name }}
{{- end -}}
{{- end -}}
```

Reply:
Should we really use `.Values.serviceAccount.create` in the templating for the name? The flag is used to determine whether or not to create the sa object at all, and imho that's it's purpose and seems unnecessary to define the name. Also the default function will not work as of now because the default value of `.Values.serviceAccount.name` in values.yaml is 'default'. We can perhaps change that to ''.

If we do that, I suppose we can simplify this and just define this as:

```
{{- define "cluster-autoscaler.serviceAccountName" -}}
{{ default (include "cluster-autoscaler.fullname" .) .Values.serviceAccount.name }}
{{- end -}}
```

Unless `cluster-autoscaler.serviceAccountName` is used elsewhere

Reply:
My first concern here is the name of the template function needs to change, since they're global.

If `.Values.serviceAccount.create=false`, then a user can specify the name of the existing ServiceAccount they want to use via `.Values.serviceAccount.name`.

What I pasted, as generated by Helm 3.1.1, gives the user the additional option of using the namespace default ServiceAccount by not specifying `.Values.serviceAccount.name`.

Reply:
hmmm, that makes sense. I hadn't really considered the need for specifying the name of an externally created SA :man_facepalming: 

I was only interested in overriding the default SA name when SA is created using the chart. But I see that ur suggestion handles all cases, provided we make `serviceAccount.name` blank in values.yaml. Is that right?

Reply:
hehe no worries. pasting from the autogenerated `values.yaml`:
```yaml
serviceAccount:
  # Specifies whether a service account should be created  create: true
  # Annotations to add to the service account
  annotations: {}
  # The name of the service account to use.
  # If not set and create is true, a name is generated using the fullname template
  name:
```

Reply:
if we went that route, we'd need to change e.g. https://github.com/helm/charts/blob/630e5b8/stable/cluster-autoscaler/templates/serviceaccount.yaml#L10 too..

Reply:
I'm not sure I follow. isn't this what the values.yaml is like now:

```yaml
rbac:
  ## If true, create & use RBAC resources
  ##
  create: false
  ## If true, create & use Pod Security Policy resources
  ## https://kubernetes.io/docs/concepts/policy/pod-security-policy/
  pspEnabled: false
  ## if rbac.create is false or (if rbac.create is true and rbac.serviceAccount.create is false)
  ## the service account rbac.serviceAccount.name will be used instead
  serviceAccount:
    create: true
    name: ""
  ## Annotations for the Service Account
  ##
  serviceAccountAnnotations: {}
```
are u suggesting that we change the structure of that a bit to make it `rbac.serviceAccount.annotations` instead of `rbac.serviceAccountAnnotations` ?

Reply:
I think it would be good, but also probably out of scope here.

Reply:
I definitely like the new proposed structure. It would I suppose also be a breaking change.

I can add the right comments on values.yaml like U suggested and mention the breaking change of sa annotations config. You reckon that should change the chart version to 7.3.0?

Reply:
I added some comments for `serviceAccount` `name` and `create` like u suggested in `values.yaml` and `README.md`. I have created a separate commit for changing `serviceAccountAnnotations` to `serviceAccount.annotations`. Let me know if it can go in the same PR or not.

Reply:
Thanks. I wonder if we should go ahead and move `serviceAccount` to the top level, instead of `rbac.serviceAccount`. Either way, we need to bump the major version now, since the annotations change breaks backward compatibility.

Reply:
I think the `rbac.serviceAccount` pattern is quite consistent with other charts I've used. Also SA is a part of rbac, right ?

Reply:
also, I've already bumped it to `7.3.0`. You aren't suggesting bumping it upto `8.0.0`, right ?

Reply:
I was suggesting to bump the major version to 8.0.0, since the change from `rbac.serviceAccountAnnotations` to `rbac.serviceAccount.annotations` breaks backward compatibility.

Instead, I think it's probably best just to remove that commit, since the modern convention is just `serviceAccount.annotations`, and renaming to something else probably doesn't do much good anyway.

Reply:
k, I'll remove that commit